### PR TITLE
Issue 9004: Pulsar Schema API: provide Type information for Fields

### LINKED ISSUE #9004: Pulsar Schema API: provide Type information for Fields
Currently the "Field" API reports only the field name and a field index.
Kafka Connects reports the data type as well.
This is very useful for  code that processes inbound events with general purpose code

**Describe the solution you'd like**
I would like to add a new property "type" (SchemaType.Type type) that reports the current type.

**Additional context**
Kafka Connect has this feature, also it looks like there is no good support for INT8, INT16, and "Map" data type.
All of these types are supported in AVRO and we should be able to use them

### PR DESCRIPTION
**Pull request by eolivelli**
Fixes #9004

### Motivation
GenericRecord API exposes the list of `Fields` but not the datatype associated to the field.
this patch introduces support for retrieving the `Schema` for each Field.

This information is available only for the limited subset of data types supported by Pulsar, for instance Avro MAPS and ARRAYS are not implemented and there is not way to represent such data types using SchemaType.
In the future when support for new datatype will be added we will support those new types in this system.  

The implementation covers Avro and JSON schema types.

With this implementation Pulsar IO support for Metadata will be more similar the one in Kafka Connect, even we are still far as we do not have full support for all AVRO datatypes. 

### Modifications
- add new Field#getSchema method
- add new version of RecordSchemaBuilder#field that accepts a Schema instead of a GenericSchema (` FieldSchemaBuilder field(String fieldName, org.apache.pulsar.client.api.Schema<?> genericSchema);`)
- add relevant test cases

### Verifying this change
This change added tests.
I tested manually that on Sinks the information is available, there is no need to add more integration tests

### Documentation
  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericJsonSchema.java [hunk 1]
```diff
@@ -43,4 +51,88 @@ public GenericJsonSchema(SchemaInfo schemaInfo) {
...
+                if (fn.isInt()) {
+                    return GenericSchema.INT32;
+                } else if (fn.isLong()) {
+                    return GenericSchema.INT64;
+                } else if (fn.isFloat()) {
+                    return GenericSchema.FLOAT;
+                } else if (fn.isDouble()) {
+                    return GenericSchema.DOUBLE;
+                } else {
+                    return null;
```
Comment (dlg99) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericJsonSchema.java:76]: (here and in other places) Consider using `Optional<>`. 
Are you using nulls for the performance reasons? how frequently this code runs?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericSchemaImpl.java [hunk 2]
```diff
@@ -41,10 +43,126 @@ protected GenericSchemaImpl(SchemaInfo schemaInfo) {
...
+            case STRING:
+                return GenericSchema.STRING;
+            case INT:
+                return GenericSchema.INT32;
+            case UNION:
+                // this is very common while representing NULLABLE types
+                // the first entry is "null", the second is the effective type
+                List<Schema> types = schema.getTypes();
+                if (types.size() == 2) {
+                    if (types.stream().anyMatch(s -> s.getType() == Schema.Type.NULL)) {
```
Comment (dlg99) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericSchemaImpl.java:78]: `if (types.size() == 2 && types.stream().filter(s -> s.getType() != Schema.Type.NULL)).count() == 1)` ?
removes need for `.orElse` and null check later

File: pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/generic/SchemaFileTypeConversionTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Test
+    public void testGenericJsonSchema() {
+        Schema<MyStruct> encodeSchema = Schema.JSON(MyStruct.class);
+        GenericSchema decodeSchema = GenericJsonSchema.of(encodeSchema.getSchemaInfo());
+        testSchema(decodeSchema, SchemaType.JSON);
+    }
+
+    private void testSchema(GenericSchema decodeSchema, SchemaType schemaTypeForStructs) {
+        findField(decodeSchema.getFields(), "theString", SchemaType.STRING);
```
Comment (dlg99) [inline pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/generic/SchemaFileTypeConversionTest.java:72]: add tests for the field that cannot be matched to anything in the schema (`"theUnknown"`)

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericSchemaImpl.java [hunk 4]
```diff
@@ -41,10 +43,126 @@ protected GenericSchemaImpl(SchemaInfo schemaInfo) {
...
+                        Schema nonNull = types.stream().filter(s -> s.getType() != Schema.Type.NULL).findFirst().orElse(null);
+                        if (nonNull != null) {
+                            return convertFieldSchema(nonNull, mainType);
+                        }
+                    }
+                }
+                return null;
+            case NULL:
```
Comment (dlg99) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericSchemaImpl.java:86]: add tests for these special cases

**Review by dlg99 — COMMENTED** [review]
nitpicks here and there

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/schema/RecordSchemaBuilder.java [hunk 5]
```diff
@@ -54,7 +54,17 @@
...
+
+    /**
+     * Add a field with the given name and genericSchema to the record.
+     *
+     * @param fieldName name of the field
+     * @param genericSchema schema of the field
+     * @return field schema builder to build the field.
+     */
+    @Deprecated
+    FieldSchemaBuilder field(String fieldName, GenericSchema<?> genericSchema);
```
Comment (sijie) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/schema/RecordSchemaBuilder.java:67]: Any reason why do you deprecated it?

**Review by sijie — CHANGES_REQUESTED** [review]
I like the idea to introduce type information for fields. However, I don't think the approach is taking the right approach. It conflates the "schema" concept in the message level with the field type. Because not all schemas can become a field in a struct. 

When we design Pulsar schema, we don't want to invent yet another type system. because all the serialization systems already have their own Type Systems. So do computation engines. If you introduce another "Pulsar" type system, you have to handle different conversations between AVRO, JSON, and PROTOBUF and with different data processing engines. The conversion process is error-prone.

We want to be stick to open standards and use the type system that each schema already has. We don't want to reinvent a wheel again. 

Hence the correct direction is just to expose whether the generic record is AVRO or JSON and use their own type systems.

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/schema/RecordSchemaBuilder.java (hunk 5 above)
Comment (eolivelli) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/schema/RecordSchemaBuilder.java:67]: I have deprecated it in favour of the new method that allows a `Schema<?>`
the new method is more general, and it should be preferred.
I cannot remove the one with `GenericSchema<?>` in order to keep binary compatibility with compiled code

**Comment by eolivelli** [comment]
> When we design Pulsar schema, we don't want to invent yet another type system.

@sijie
In Kafka Connect we see this great abstraction of [Schema](https://kafka.apache.org/11/javadoc/org/apache/kafka/connect/data/Schema.html)
and this makes it very easy to implement Connectors (especially Sinks) in a way that is not dependant on the serialization mechanism.

Therefore in Pulsar IO we do not allow the function implementation to access the underlying Avro libraries, I like this fact and I would like not to change it.

So my thinking is that we need our way to describe a Schema with a Pulsar API.
I would like also to follow up in order to add to the Pulsar Schema API all of the data types that are missing, like Arrays and Maps.

In my opinion without these features Pulsar IO won't be able to compete with Kafka Connect, and also to "port" existing Kafka Connect workflows to the Pulsar Ecosystem

**Comment by sijie** [comment]
@eolivelli 

There are two questions here. Let's not couple them together.

1) Whether we want to introduce a type system or not? If we want to introduce a type system what would be the approach.
2) How connector can access the schema information?

For the first one, I have a really strong different opinion on introducing a type system. I have seen problems with converting types between different systems. AVRO is already an open-standard that all the existing computing engines support. If we are introducing another type of system, we introduce a lot of troubles when integrating with other computing engines. That's something I would avoid. If you want to do that, let's think carefully. It is not just about connectors. Pulsar has a broader set of integrations that relies on schema. We can't afford to maintain different type systems and different approaches for different integrations. You end up need to maintain a lot of converters between Pulsar schemas with many other systems. It is going to be a nightmare. 

For the second one, we should just let users access the underlying libraries. AVRO, JSON, and PROTOBUF have different type systems. Let's use their existing tools and not reinvent one.

> without these features Pulsar IO won't be able to compete with Kafka Connect, and also to "port" existing Kafka Connect workflows to the Pulsar Ecosystem

First of all, Pulsar I/O doesn't compete with Kafka Connect. Pulsar I/O is an extension of the Pulsar ecosystem. Kafka Connect is the extension to Kafka. They might share similar goals. But they don't compete with each other. 

Secondly, we embrace the Kafka ecosystem by providing a fully compatible wrapper to run existing Kafka connectors without introducing a type system. The key for this to work is using an open standard like ARVO. 

```
Kafka Schema -> [ Open Standards: AVRO ] -> Pulsar Schema 
```

**Comment by eolivelli** [comment]
Sorry, I didn't mean that Pulsar IO is a competitor for Kafka Connect, but I would like to share my experience in porting projects that are working on Kafka Connect to Pulsar IO Sinks.

I recently got stuck in this problem, that when you deal with GenericRecord you cannot have metadata about the Schema, you can only access the raw schema bytes or you have to use Java reflection in order to deal with the type of each field (and this also is a problem because for null values you do not know the original datatype).

GenericRecord is a wonderful abstraction over generic data structures and it looks to me that the most missing part of the story is about having this kind of metadata about the structure.

GenericRecord already provides many APIs to create records and create schemas (RecordSchemaBuilder and GenericRecordBuilder) , the only missing part here is to "read" that metadata.

Once we have this API working with structured data with Pulsar IO (and probably with PulsarFunctions in general) will be more easier.

Dealing directly with Avro (and Protobuf...) is not a good way to go for my usecases, because I will have to explicitly write code for every schema type, and also enter the details of every technology. 

For simple cases (but GenericRecord is not that simple, it is already very powerful, as it is already able to deal with nested structs) it is a great feature to be able to code your Pulsar Sink independently from the physical encoding of the records, the same way we do with Compression.


cc @shiv4289 @aahmed-se

**Comment by eolivelli** [comment]
@sijie I sent a new PR that allows you to access the underlying Schema information

https://github.com/apache/pulsar/pull/9614

I am going to close this PR if you are fine with that new approach

**Comment by eolivelli** [comment]
closing in favour of #9614

**Comment by vroyer** [comment]
By making this decision (no generic schema on top of multiple encoding types), the Pulsar IO ecosystem won't be able to do data transformation without dealing with the data serialisation/encoding. Don't you think it is a limitation compared to the Kafka connect ecosystem ?

**Pull request closed without merging**